lang/cs/Org.Apache.REEF.Network/Group/Task/Impl/OperatorTopology.cs

```diff
@@ -322,27 +305,7 @@ public T ReceiveFromChildren(IReduceFunction<T> reduceFunction)
                 throw new ArgumentNullException("reduceFunction");
             }
 
-            var receivedData = new List<T>();
-            var childrenToReceiveFrom = new HashSet<string>(_children.Select(node => node.Identifier));
-
-            while (childrenToReceiveFrom.Count > 0)
-            {
-                var childrenWithData = GetNodeWithData(childrenToReceiveFrom);
-
-                foreach (var child in childrenWithData)
-                {
-                    T[] data = ReceiveFromNode(child);
-                    if (data == null || data.Length != 1)
-                    {
-                        throw new InvalidOperationException("Received invalid data from child with id: " + child.Identifier);
-                    }
-
-                    receivedData.Add(data[0]);
-                    childrenToReceiveFrom.Remove(child.Identifier);
-                }
-            }
-
-            return reduceFunction.Reduce(receivedData);
+            return reduceFunction.Reduce(_childNodeContainer.GetDataFromAllChildren());
```

Comment:
This can potentially significantly increase the memory footprint, right? Consider the case where we are not using pipelining and T is a large vector that we want to reduce. We now wait for all of them to be present before doing reduce. For flat topology with P nodes there will be P such large vectors, although one can argue that we should use binary topology then :) . Since we in any case are waiting for all children and reduce is assumed to be associative and commutative, I am wondering if we should apply reduce incrementally one child at a time. This can potentially lead to smaller memory footprints. Does that make sense?